concordium-contracts-common-derive/src/lib.rs

```diff
@@ -348,155 +81,567 @@ fn impl_deserial(ast: &syn::DeriveInput) -> syn::Result<TokenStream> {
 #[proc_macro_derive(Serial, attributes(concordium))]
 pub fn serial_derive(input: TokenStream) -> TokenStream {
     let ast = parse_macro_input!(input);
-    unwrap_or_report(impl_serial(&ast))
+    unwrap_or_report(derive::impl_serial(&ast))
 }
 
-fn impl_serial_field(
-    field: &syn::Field,
-    ident: &proc_macro2::TokenStream,
-    out: &syn::Ident,
-) -> syn::Result<proc_macro2::TokenStream> {
-    let root = get_root();
-
-    if let Some(size_length) = find_length_attribute(&field.attrs)? {
-        let l = format_ident!("U{}", 8 * size_length);
-        Ok(quote!({
-            #root::SerialCtx::serial_ctx(#ident, #root::schema::SizeLength::#l, #out)?;
-        }))
-    } else {
-        Ok(quote! {
-            #root::Serial::serial(#ident, #out)?;
-        })
-    }
+/// A helper macro to derive both the Serial and Deserial traits.
+/// `[derive(Serialize)]` is equivalent to `[derive(Serial, Deserial)]`, see
+/// documentation of the latter two for details and options:
+/// [`derive(Serial)`](./derive.Serial.html),
+/// [`derive(Deserial)`](./derive.Deserial.html).
+#[proc_macro_derive(Serialize, attributes(concordium))]
+pub fn serialize_derive(input: TokenStream) -> TokenStream {
+    unwrap_or_report(serialize_derive_worker(input))
 }
 
-fn impl_serial(ast: &syn::DeriveInput) -> syn::Result<TokenStream> {
-    let data_name = &ast.ident;
+fn serialize_derive_worker(input: TokenStream) -> syn::Result<TokenStream> {
+    let ast = syn::parse(input)?;
+    let mut tokens = derive::impl_deserial(&ast)?;
+    tokens.extend(derive::impl_serial(&ast)?);
+    Ok(tokens)
+}
 
-    let span = ast.span();
+/// Derive the DeserialWithState trait. See the documentation of
+/// [`derive(Deserial)`](./derive.Deserial.html) for details and limitations.
+///
+/// This trait should be derived for `struct`s or `enum`s that have fields with
+/// [`StateBox`](../concordium_std/struct.StateBox.html),
+/// [`StateSet`](../concordium_std/struct.StateSet.html), or
+/// [`StateMap`](../concordium_std/struct.StateMap.html). Please note that it is
+/// necessary to specify the generic parameter name for the
+/// [`HasStateApi`](../concordium_std/trait.HasStateApi.html) generic parameter.
```

Comment:
I don't think these links work after the move. We should change them to the full URL of concordium-std on docs.rs

Reply:
They appear multiple places.

Reply:
So these links actually work, as long as you read it from the docs page of `concordium-std`.
The documentation for `concordium-contracts-common` only includes them if the `smart-contract` feature is enabled, which is not the case when releasing to docs.rs.
However, they do appear broken in the standalone docs for `concordium-contracts-common-derive`.

I would prefer not to have full URLs, since it won't navigate to the matching version of the documentation for these types.
So I would say we keep it as is.

Reply:
I see. Ok, that is fair. Let's keep them.